flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/TwoInputStreamTaskTest.java

```diff
@@ -671,5 +737,100 @@ public InputSelection nextSelection() {
 			return InputSelection.ALL;
 		}
 	}
+
+	/**
+	 * Uses to test handling the EndOfInput notification.
+	 */
+	private static class TestBoundedTwoInputOperator extends AbstractStreamOperator<String>
+		implements TwoInputStreamOperator<String, String, String>, BoundedMultiInput {
+
+		private static final long serialVersionUID = 1L;
+
+		private final String name;
+
+		private volatile int outputRecords1;
+		private volatile int outputRecords2;
+
+		private static final CompletableFuture<Void> success = CompletableFuture.completedFuture(null);
+		private static final CompletableFuture<Void> failure = FutureUtils.completedExceptionally(
+			new Exception("Not in line with expectations."));
+
+		public TestBoundedTwoInputOperator(String name) {
+			this.name = name;
+		}
+
+		@Override
+		public void processElement1(StreamRecord<String> element) {
+			output.collect(element.replace("[" + name + "-1]: " + element.getValue()));
+			outputRecords1++;
+
+			synchronized (this) {
+				this.notifyAll();
+			}
+		}
+
+		@Override
+		public void processElement2(StreamRecord<String> element) {
+			output.collect(element.replace("[" + name + "-2]: " + element.getValue()));
+			outputRecords2++;
+
+			synchronized (this) {
+				this.notifyAll();
+			}
+		}
+
+		@Override
+		public void endInput(int inputId) {
+			output.collect(new StreamRecord<>("[" + name + "-" + inputId + "]: Bye"));
+
+			if (inputId == 1) {
+				outputRecords1++;
+			} else {
+				outputRecords2++;
+			}
+
+			synchronized (this) {
+				this.notifyAll();
+			}
+		}
+
+		public void waitOutputRecords(int inputId, int expectedRecords) throws Exception {
```

Comment:
ditto: why do you need this method?

Reply:
This ensures that the output is strictly ordered, to verify whether `endInput()` is called immediately when `input1` arrives the end.

Reply:
Is there a reason why you can not use `testHarness.waitForInputProcessing`?

Reply:
You're right, `testHarness.waitForInputProcessing` works well. I was misled by its name before.